src/eval.ml

```diff
@@ -218,23 +136,145 @@ and eq_modulo : term -> term -> bool = fun a b ->
+and tree_walk : dtree -> stack -> (term * stack) option = fun tree stk ->
+  let (lazy capacity, lazy tree) = tree in
+  let open Tree_types in
+  let open Dtree in
+  let stk = Stack.of_list stk in
+  let vars = Array.make capacity Kind in (* dummy terms *)
+  let boundv = Array.make capacity TE_None in
+  (* [walk t s c m] where [s] is the stack of terms to match and [c] the
+     cursor indicating where to write in the [env] array described in
+     {!module:Terms} as the environment of the RHS during matching.  [m]
+     maps the free variables contained in the tree to the free variables
+     used in this evaluation. *)
+  let rec walk tree stk cursor fresh_vars =
+    match tree with
+    | Fail                                                -> None
+    | Leaf(env_builder, act)                              ->
+        (* Allocate an environment for the action. *)
+        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+        (* Retrieve terms needed in the action from the [vars] array. *)
+        let fn (pos, (slot, xs)) =
+          match boundv.(pos) with
+          | TE_Vari(_) -> assert false
+          | TE_Some(_) -> env.(slot) <- boundv.(pos)
+          | TE_None    ->
+              if Array.length xs = 0 then
+                let t = unfold vars.(pos) in
+                let b = Bindlib.raw_mbinder [||] [||] 0 mkfree (fun _ -> t) in
+                env.(slot) <- TE_Some(b)
+              else
+                let b = lift vars.(pos) in
+                let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+                env.(slot) <- TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs b))
+        in
+        List.iter fn env_builder;
+        (* Actually perform the action. *)
+        Some(Bindlib.msubst act env, Stack.to_list stk)
+    | Condition({ ok ; condition ; fail })                ->
+        let next =
+          match condition with
+          | TcstrEq(i, j)        ->
+              if eq_modulo vars.(i) vars.(j) then ok else fail
+          | TcstrFreeVars(xs, i) ->
+              let xs = Array.map (fun e -> VarMap.find e fresh_vars) xs in
+              (* We first attempt to match [vars.(i)] directly. *)
+              let b = Bindlib.bind_mvar xs (lift vars.(i)) in
+              let fn _ x = not (Bindlib.occur x b) in
+              let r = VarMap.for_all fn fresh_vars in
+              if r then (boundv.(i) <- TE_Some(Bindlib.unbox b); ok) else
+              (* As a last resort we try matching the SNF. *)
+              let b = Bindlib.bind_mvar xs (lift (snf vars.(i))) in
+              let fn _ x = not (Bindlib.occur x b) in
+              let r = VarMap.for_all fn fresh_vars in
+              if r then (boundv.(i) <- TE_Some(Bindlib.unbox b); ok) else fail
+        in
+        walk next stk cursor fresh_vars
+    | Node({swap; children; store; abstraction; default}) ->
+        try
+          let (left, examined, right) = Stack.destruct stk swap in
```

Comment:
Here I think I can wrap the `Stack.destruct` alone in a `try ... with Not_found ->`, does that look good to you @rlepigre ?

Reply:
I don't think you'll be able to do that here.

Reply:
```
let o = try Stack.destruct stk swap
        with Not_found -> None in
match o with None -> None
           | Some(a, b, c) -> ...
```
?

Reply:
In that case, it is better to change the type of `Stack.destruct` to return an option.

Reply:
Right